# Ejercicios resueltos — análisis de una imagen con NumPy

La imagen `beagle.jpg` se trabaja como un arreglo tridimensional: **alto × ancho × canales RGB**.

In [1]:
import numpy as np
import matplotlib.pyplot as plt

img = plt.imread('beagle.jpg')
img.shape

(1200, 1106, 3)

**Respuesta:** la imagen tiene 1200 píxeles de alto, 1106 píxeles de ancho y 3 canales de color: rojo, verde y azul (RGB).

In [2]:
print("Dimensiones:", img.ndim)
print("Shape:", img.shape)
print("Tipo de dato:", img.dtype)

Dimensiones: 3
Shape: (1200, 1106, 3)
Tipo de dato: uint8


## Ejercicio 1: inspeccionar un píxel

In [3]:
pixel = img[10, 10]
print("RGB del píxel [10,10]:", pixel)

RGB del píxel [10,10]: [22 35  5]


El píxel `[10,10]` contiene tres intensidades: **R=22, G=35 y B=5**. Cada valor representa la intensidad de uno de los canales RGB.

## Ejercicio 2: separar los canales RGB

In [4]:
rojo = img[:, :, 0]
verde = img[:, :, 1]
azul = img[:, :, 2]

print("Canal rojo:", rojo.shape)
print("Canal verde:", verde.shape)
print("Canal azul:", azul.shape)

Canal rojo: (1200, 1106)
Canal verde: (1200, 1106)
Canal azul: (1200, 1106)


## Ejercicio 3: crear una versión en escala de grises

In [5]:
gris = img.mean(axis=2)
print("Shape de la imagen en gris:", gris.shape)
print("Dimensiones:", gris.ndim)

Shape de la imagen en gris: (1200, 1106)
Dimensiones: 2


Al promediar los tres canales RGB, cada píxel queda representado por un solo valor. Por eso la imagen resultante tiene dos dimensiones.

## Ejercicio 4: invertir los colores

In [6]:
img_invertida = 255 - img
print("Shape de la imagen invertida:", img_invertida.shape)
print("Tipo:", img_invertida.dtype)

Shape de la imagen invertida: (1200, 1106, 3)
Tipo: uint8


## Ejercicio 5: recortar una región de la imagen

In [7]:
recorte = img[200:600, 300:800]
print("Shape del recorte:", recorte.shape)

Shape del recorte: (400, 500, 3)


## Visualización de los resultados

In [8]:
# Ejecuta esta celda en Jupyter para comparar los resultados.
plt.figure(figsize=(6, 4))
plt.imshow(img)
plt.axis('off')
plt.title('Imagen original')
plt.show()

plt.figure(figsize=(6, 4))
plt.imshow(gris, cmap='gray')
plt.axis('off')
plt.title('Escala de grises')
plt.show()

plt.figure(figsize=(6, 4))
plt.imshow(img_invertida)
plt.axis('off')
plt.title('Colores invertidos')
plt.show()

## Conclusión

El ejercicio muestra que una imagen digital puede manipularse como un `ndarray`. La indexación permite consultar píxeles y recortar regiones, mientras que las operaciones vectorizadas permiten transformar todos los píxeles de manera sencilla.